### ISSUE TITLE
`helm uninstall` with `--keep-history` does not suspend previous deployed releases

### ISSUE THREAD

**Issue by hiddeco** [helm/helm#12556]
Assuming the following scenario:

```console
$ helm install podinfo podinfo/podinfo
NAME: podinfo
LAST DEPLOYED: Fri Nov 10 15:07:18 2023
NAMESPACE: helm-system
STATUS: deployed
REVISION: 1
NOTES:
1. Get the application URL by running these commands:
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl -n helm-system port-forward deploy/podinfo 8080:9898

$ helm upgrade podinfo podinfo/podinfo --set replicaCount=2 --set faults.unready=true --wait --timeout=10s
Error: UPGRADE FAILED: timed out waiting for the condition

$ helm uninstall podinfo --keep-history
release "podinfo" uninstalled
```

I would have expected the output from `helm history` to equal:

```console
$ helm history podinfo
REVISION	UPDATED                 	STATUS     	CHART        	APP VERSION	DESCRIPTION            
1       	Fri Nov 10 15:07:18 2023	superseded   	podinfo-6.5.0	6.5.0      	Install complete       
2       	Fri Nov 10 15:09:10 2023	uninstalled	podinfo-6.5.0	6.5.0      	Uninstallation complete
```

However, the actual output is:

```console
$ helm history podinfo
REVISION	UPDATED                 	STATUS     	CHART        	APP VERSION	DESCRIPTION            
1       	Fri Nov 10 15:07:18 2023	deployed   	podinfo-6.5.0	6.5.0      	Install complete       
2       	Fri Nov 10 15:09:10 2023	uninstalled	podinfo-6.5.0	6.5.0      	Uninstallation complete
```

Which actually can result in surprising behavior when performing a `helm upgrade` again with the `--atomic` flag set:

```console
$ helm upgrade podinfo podinfo/podinfo --set replicaCount=2 --set faults.unready=true --wait --timeout=10s --atomic
Error: UPGRADE FAILED: release podinfo failed, and has been rolled back due to atomic being set: timed out waiting for the condition

$ helm history podinfo
REVISION	UPDATED                 	STATUS     	CHART        	APP VERSION	DESCRIPTION                                                  
1       	Fri Nov 10 15:07:18 2023	superseded 	podinfo-6.5.0	6.5.0      	Install complete                                             
2       	Fri Nov 10 15:09:10 2023	uninstalled	podinfo-6.5.0	6.5.0      	Uninstallation complete                                      
3       	Fri Nov 10 15:16:54 2023	failed     	podinfo-6.5.0	6.5.0      	Upgrade "podinfo" failed: timed out waiting for the condition
4       	Fri Nov 10 15:17:05 2023	deployed   	podinfo-6.5.0	6.5.0      	Rollback to 1  
```

---

I think this can either be solved by implementing a likewise fix as was done for `helm rollback` (see #2941), or by superseding `deployed` releases as soon as a mutation to the cluster has been made. Which seems to be more correct (and solving the issue at the root of the problem) in general.

**Comment by mattfarina** [comment]
We're going to have to look more into the history on this one. For example, if revision 1 was _superseded_ than the last (atomic) upgrade would fail no matter what. It would have not previous deployed version to be based on. But, are upgrades out of a failure supposed to go forward based on the last successful deployment or the last deployment?

**Comment by mattfarina** [comment]
Note, in the example here... when the 2nd `helm upgrade` is run that fails to upgrade, this example has the first revision running successfully with one pod while at the same time the second revision (that's intentionally failing) is starting but has no ready pods. It's caught half way.

**Comment by hiddeco** [comment]
> But, are upgrades out of a failure supposed to go forward based on the last successful deployment or the last deployment?

I would think that given the fact that a successful `helm upgrade` before a `helm uninstall --keep-history` does not succeed in a successful `helm upgrade`, the same behavior should happen when a `helm uninstall --keep-history` is run after a failed upgrade. Which at present, appears to not happen due to the release being left in a `deployed` state.

In addition, I would argue that even besides this (most probably unintended) side-effect (which I mainly discovered due to trying to see if this behavior could result in other unexpected behavior), keeping the release in a `deployed` state after an uninstall is wrong (as the resources have been removed from the cluster, the history just has not).

**Comment by mattfarina** [comment]
> keeping the release in a deployed state after an uninstall is wrong (as the resources have been removed from the cluster, the history just has not).

Good point. It really depends on where the failure is in the process. In this case, the failure is after the resources in the cluster have already been updated.

**Comment by hiddeco** [comment]
> It really depends on where the failure is in the process.

This I can understand, but the current behavior where a deployed and failed release exist when a post-upgrade hook fails is an example where I would expect the existing release to be marked as superseded.

The same goes for failures which are due to a timeout waiting for resources to reach their ready state, as at this point none of the previous resource data should be around.
